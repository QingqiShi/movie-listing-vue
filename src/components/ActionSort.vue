<template>
  <div class="sort">
    <span class="sort__label">Sort By</span>
    <template v-for="(link, i) of links">
      <a
        :key="`sort-link-${link.property}`"
        :class="link.class"
        href="#"
        @click="link.handleClick"
      >
        {{ link.label }}
        <span v-if="link.active">{{ sortOrder ? "↑" : "↓" }}</span>
      </a>
      <span v-if="i == 0" :key="`sort-${i}`" class="sort__seperator">|</span>
    </template>
  </div>
</template>

<script>
import { mapState } from "vuex";

export default {
  computed: {
    ...mapState({
      sortOrder: state => state.movies.sortOrder,
      sortProperty: state => state.movies.sortProperty
    }),
    links() {
      const popularityActive = this.sortProperty === "popularity";
      const popularityLink = {
        property: "popularity",
        label: "Popularity",
        active: popularityActive,
        class: {
          sort__link: true,
          "sort__link--popularity": true,
          "sort__link--active": popularityActive
        },
        handleClick: () => this.updateSort("popularity")
      };

      const ratingActive = this.sortProperty === "vote_average";
      const ratingLink = {
        property: "vote_average",
        label: "Rating",
        active: ratingActive,
        class: {
          sort__link: true,
          "sort__link--vote_average": true,
          "sort__link--active": ratingActive
        },
        handleClick: () => this.updateSort("vote_average")
      };

      return [popularityLink, ratingLink];
    }
  },
  methods: {
    updateSort(property) {
      const newOrder = this.sortProperty === property ? !this.sortOrder : false;
      this.$store.commit("setSort", {
        property,
        order: newOrder
      });
      this.$router.push({
        query: {
          ...this.$router.currentRoute.query,
          sortProperty: property,
          sortOrder: newOrder
        }
      });
    }
  }
};
</script>

<style lang="scss">
.sort {
  height: 100%;
  display: flex;
  align-items: center;
  margin-right: 1rem;

  &__link {
    color: $accent;
    font-family: $font-heading;
    text-decoration: none;
    border-radius: 0.3rem;

    &:hover {
      color: $on-primary;
    }

    &--active {
      padding: 0.4rem 0.8rem;
      color: $on-primary;
      background-color: $accent;

      &:hover {
        color: $accent;
        background-color: $on-primary;
      }
    }
  }

  &__label {
    margin-right: 0.5rem;
  }

  &__seperator {
    margin: 0 0.5rem;
  }
}
</style>
