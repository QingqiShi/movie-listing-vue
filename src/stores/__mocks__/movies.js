const state = {
  configuration: {
    imageBaseUrl: "https://images.com/",
    posterSizes: ["w100", "w300", "original"],
    backdropSizes: ["w100", "w300", "original"]
  },
  movies: [
    {
      id: 1000,
      title: "Avatar",
      genre_ids: [0],
      popularity: 230,
      vote_average: 5
    }
  ],
  genres: [{ id: 0, name: "Action" }, { id: 1, name: "Horror" }],
  sortOrder: false,
  sortProperty: "popularity",
  filterRating: 3,
  filterGenres: { 0: false, 1: true },
  error: ""
};

const getters = {
  /** @param {typeof state} state */
  ratingFilteredMovies: state => state.movies,
  availableGenres: () => [{ id: 0, name: "Action" }],
  filteredMovies: () => []
};

const mutations = {
  setConfiguration: jest.fn(),
  setMovies: jest.fn(),
  setGenres: jest.fn(),
  setSort: jest.fn(),
  setFilter: jest.fn(),
  setError: jest.fn()
};

const actions = {
  initialFetch: jest.fn(() => Promise.resolve()),
  fetchConfiguration: jest.fn(() => Promise.resolve()),
  fetchMovies: jest.fn(() => Promise.resolve()),
  fetchGenres: jest.fn(() => Promise.resolve())
};

export default { state, getters, mutations, actions };
