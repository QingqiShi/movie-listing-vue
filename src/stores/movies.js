import api from "@/constants/api";
import app from "@/constants/app";

const { CONFIGURATION, NOW_PLAYING, GENRES_LIST } = api;
const { LOCALE } = app;

const state = {
  configuration: {},
  /** @type {Movie[]} List of all movies */
  movies: [],
  /** @type {Genre[]} List of all genres */
  genres: [],
  /** @type {boolean} True for accending, false for decending */
  sortOrder: false,
  /** @type {"popularity"|"vote_average"} */
  sortProperty: "popularity",
  /** @type {number} Number between 0 and 10 */
  filterRating: 3,
  /** @type {{[id:number]: boolean}} */
  filterGenres: {},
  /** @type {string} */
  error: ""
};

const getters = {
  /** @type {(state:State) => Movie[]} */
  ratingFilteredMovies: state =>
    state.movies.filter(movie => movie.vote_average >= state.filterRating),

  /** @type {(state:State, getters:Getters) => Genre[]} */
  availableGenres: (state, getters) =>
    state.genres.filter(genre => {
      for (let movie of getters.ratingFilteredMovies) {
        if (movie.genre_ids.includes(genre.id)) {
          return true;
        }
      }
      return state.filterGenres[genre.id];
    }),

  /** @type {(state:State, getters:Getters) => Movie[]} */
  filteredMovies: (state, getters) => {
    const { availableGenres, ratingFilteredMovies } = getters;
    const { filterGenres } = state;

    const genres = availableGenres.filter(genre => filterGenres[genre.id]);
    if (!genres.length) return ratingFilteredMovies;

    return ratingFilteredMovies.filter(({ genre_ids }) =>
      genres.reduce((show, genre) => show && genre_ids.includes(genre.id), true)
    );
  }
};

const mutations = {
  /** @type {(state:State, payload:Configuration) => void} */
  setConfiguration(state, { imageBaseUrl, posterSizes, backdropSizes }) {
    state.configuration = { imageBaseUrl, posterSizes, backdropSizes };
  },

  /** @type {(state:State, movies:Movie[]) => void} */
  setMovies(state, movies) {
    state.movies = movies;
  },

  /** @type {(state:State, genres:Genre[]) => void} */
  setGenres(state, genres) {
    state.genres = genres;
    state.filterGenres = {};
    genres.forEach(genre => {
      state.filterGenres[genre.id] = false;
    });
  },

  /** @type {(state:State, payload:{property:SortProperty, order:boolean}) => void} */
  setSort(state, { property, order }) {
    state.sortOrder = order;
    state.sortProperty = property;
    state.movies = state.movies.sort((a, b) =>
      order ? a[property] - b[property] : b[property] - a[property]
    );
  },

  /** @type {(state:State, payload:{type:"rating"|"genres", value:number}) => void} */
  setFilter(state, { type, value }) {
    if (type === "rating") {
      state.filterRating = value;
    } else {
      const newFilters = { ...state.filterGenres };
      newFilters[value] = !state.filterGenres[value];
      state.filterGenres = newFilters;
    }
  },

  /** @type {(state:State, errorMsg:string) => void} */
  setError(state, errorMsg) {
    state.error = errorMsg;
  }
};

const actions = {
  /** @param {import('vuex').ActionContext} context */
  async initialFetch({ commit, dispatch }) {
    try {
      await Promise.all([
        dispatch("fetchConfiguration"),
        dispatch("fetchMovies"),
        dispatch("fetchGenres")
      ]);
      commit("setSort", { property: "popularity", order: false });
    } catch (e) {
      commit("setError", e);
    }
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchConfiguration({ commit }) {
    const result = await fetchApi(
      CONFIGURATION(),
      "FetchConfiguration",
      result => result && result.images
    );

    commit("setConfiguration", {
      imageBaseUrl: result.images.base_url,
      posterSizes: result.images.poster_sizes,
      backdropSizes: result.images.backdrop_sizes
    });
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchMovies({ commit }) {
    const result = await fetchApi(
      NOW_PLAYING(LOCALE),
      "FetchMovies",
      result => result && result.results
    );

    commit("setMovies", result.results);
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchGenres({ commit }) {
    const result = await fetchApi(
      GENRES_LIST(LOCALE),
      "FetchGenres",
      result => result && result.genres
    );

    commit("setGenres", result.genres);
  }
};

/**
 * @param {string} url
 * @param {string} name
 */
async function fetchApi(url, name, validator) {
  let result;

  try {
    const response = await fetch(url);
    result = await response.json();
  } catch {
    throw `${name} failed, check your network and try again`;
  }

  if (!validator(result)) {
    throw `${name} returned invalid result`;
  }

  return result;
}

export default { state, getters, mutations, actions };

/* Type definitions to help vscode do auto suggest */

/** @typedef {{id:number, name:string}} Genre */
/**
 * @typedef {object} Movie
 * @property {number} Movie.id
 * @property {string} Movie.title
 * @property {number} Movie.popularity
 * @property {number} Movie.vote_average
 * @property {number[]} Movie.genre_ids
 * @property {string} Movie.poster_path
 */
/**
 * @typedef {object} Configuration
 * @property {string} Configuration.imageBaseUrl
 * @property {string} Configuration.posterSizes
 * @property {string} Configuration.backdropSizes
 */
/** @typedef {typeof state} State */
/**
 * @typedef {object} Getters
 * @property {Movie[]} Getters.ratingFilteredMovies
 * @property {Genre[]} Getters.availableGenres
 * @property {Movie[]} Getters.filteredMovies
 */
/** @typedef {typeof state.sortProperty} SortProperty */
