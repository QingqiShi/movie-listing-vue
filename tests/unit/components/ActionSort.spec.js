import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import ActionSort from "@/components/ActionSort.vue";
import movies from "@/stores/movies";

jest.mock("@/stores/movies");

const localVue = createLocalVue();
localVue.use(Vuex);

const routerMock = { push: jest.fn(), currentRoute: { query: { fake: true } } };

describe("ActionSort.vue", () => {
  it.each`
    property          | order
    ${"popularity"}   | ${true}
    ${"popularity"}   | ${false}
    ${"vote_average"} | ${true}
    ${"vote_average"} | ${false}
  `("renders sort by $property order $order", ({ property, order }) => {
    movies.state.sortProperty = property;
    movies.state.sortOrder = order;
    const store = new Vuex.Store({ modules: { movies } });
    const wrapper = shallowMount(ActionSort, {
      store,
      localVue,
      mocks: { $router: routerMock }
    });

    expect(wrapper).toMatchSnapshot();
  });

  describe("update sort", () => {
    let wrapper;
    const sortLinks = {};

    beforeAll(() => {
      const store = new Vuex.Store({ modules: { movies } });
      wrapper = shallowMount(ActionSort, {
        store,
        localVue,
        mocks: { $router: routerMock }
      });

      sortLinks.popularity = wrapper.find(".sort__link--popularity");
      sortLinks.rating = wrapper.find(".sort__link--vote_average");
    });

    it.each`
      startProperty     | startOrder | toClick         | expectedProperty  | expectedOrder
      ${"popularity"}   | ${false}   | ${"popularity"} | ${"popularity"}   | ${true}
      ${"popularity"}   | ${true}    | ${"rating"}     | ${"vote_average"} | ${false}
      ${"vote_average"} | ${false}   | ${"rating"}     | ${"vote_average"} | ${true}
      ${"vote_average"} | ${true}    | ${"popularity"} | ${"popularity"}   | ${false}
    `(
      "click on $toClick link (from $startProperty, $startOrder)",
      ({
        startProperty,
        startOrder,
        toClick,
        expectedProperty,
        expectedOrder
      }) => {
        movies.state.sortProperty = startProperty;
        movies.state.sortOrder = startOrder;

        sortLinks[toClick].trigger("click");
        expect(movies.mutations.setSort).toHaveBeenLastCalledWith(
          expect.anything(),
          {
            property: expectedProperty,
            order: expectedOrder
          }
        );

        expect(routerMock.push).toHaveBeenLastCalledWith({
          query: {
            fake: true,
            sortProperty: expectedProperty,
            sortOrder: expectedOrder
          }
        });
      }
    );
  });
});
