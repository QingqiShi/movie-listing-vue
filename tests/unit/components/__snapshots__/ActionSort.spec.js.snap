// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSort.vue renders sort by "popularity" order false 1`] = `
<div class="sort"><span class="sort__label">Sort By</span> <a href="#" class="sort__link sort__link--popularity sort__link--active">
    Popularity
    <span>↓</span></a> <span class="sort__seperator">|</span><a href="#" class="sort__link sort__link--vote_average">
    Rating
    <!----></a>
  <!---->
</div>
`;

exports[`ActionSort.vue renders sort by "popularity" order true 1`] = `
<div class="sort"><span class="sort__label">Sort By</span> <a href="#" class="sort__link sort__link--popularity sort__link--active">
    Popularity
    <span>↑</span></a> <span class="sort__seperator">|</span><a href="#" class="sort__link sort__link--vote_average">
    Rating
    <!----></a>
  <!---->
</div>
`;

exports[`ActionSort.vue renders sort by "vote_average" order false 1`] = `
<div class="sort"><span class="sort__label">Sort By</span> <a href="#" class="sort__link sort__link--popularity">
    Popularity
    <!----></a> <span class="sort__seperator">|</span><a href="#" class="sort__link sort__link--vote_average sort__link--active">
    Rating
    <span>↓</span></a>
  <!---->
</div>
`;

exports[`ActionSort.vue renders sort by "vote_average" order true 1`] = `
<div class="sort"><span class="sort__label">Sort By</span> <a href="#" class="sort__link sort__link--popularity">
    Popularity
    <!----></a> <span class="sort__seperator">|</span><a href="#" class="sort__link sort__link--vote_average sort__link--active">
    Rating
    <span>↑</span></a>
  <!---->
</div>
`;
